File: Sources/xct/asset.swift
//
//  asset.swift
//  xct
//
//  Created by TBXark on 2020/11/12.
//

import Foundation
import JsonMapper
import PathKit

struct RenameAsset: CommandService {
    let key: String = "rename-asset"
    let help = """
    xct rename-asset <location>
    // example: xct rename-asset ./xctdemo/Sources
    arguments:
        <location>: path to target directory
    """

    func run(arguments: [String]) {
        do {
            guard let dir = arguments.first.map({ Path($0).absolute() }) else {
                throw xctError(reason: "target location path not found")
            }
            let paths = findAllDirectoryPaths(dir, suffix: ".imageset")
            fputs("Find \(paths.count) assets in \(dir)", stdout)
            for asset in paths {
                let jsonPath = asset + "Contents.json"

                let data: Data = try jsonPath.read()
                var json: String = try jsonPath.read()
                let info = try JSONDecoder().decode(JSONElement.self, from: data)

                let assetName = asset.lastComponentWithoutExtension
                var didUpdate = false

                for image in info.images.arrayValue ?? [] {
                    guard let name = image.filename.stringValue,
                          let type = name.split(separator: ".").last else {
                        continue
                    }
                    let targetName = "\(assetName)\(image.scale.stringValue.map({ "@\($0)" }) ?? "").\(type)"
                    if targetName != name {
                        try (asset + name).move(asset + targetName)
                        didUpdate = true
                        json = json.replacingOccurrences(of: "\"filename\" *: *\"\(name)\"", with: "\"filename\" : \"\(targetName)\"", options: .regularExpression, range: nil)
                    }
                }

                if didUpdate, let jsonData = json.data(using: .utf8) {
                    try jsonPath.write(jsonData)
                }
            }
        } catch {
            fputs("error: \(error.localizedDescription)\n\(help)", stderr)
            exit(1)
        }
    }
}

struct AssetCleaner: CommandService {
    let key: String = "clean-asset"
    let help = """
    xct clean-asset <location>
    // example: xct clean-asset ./xctdemo/Sources
    arguments:
        <location>: path to target directory
    """
    func run(arguments: [String]) {
        do {
            guard let dir = arguments.first.map({ Path($0).absolute()  }) else {
                throw xctError(reason: "target location path not found")
            }
            var assets = findAllDirectoryPaths(dir, suffix: ".xcassets")
                .map({
                    (xcassets: $0, imageset: Set(findAllDirectoryPaths($0, suffix: ".imageset")))
                })
            let swiftFile = findAllFilePaths(dir, suffix: ".swift").filter({ $0.lastComponent != "R.generated.swift" })
            for file in swiftFile {
                autoreleasepool { () -> Void in
                    do {
                        let text: String = try file.read()
                        for x in 0..<assets.count {
                            var find = [Path]()
                            for asset in assets[x].imageset {
                                let name = asset.lastComponentWithoutExtension
                                if text.contains("R.image.\(name)") || text.contains("#imageLiteral(resourceName: \"\(name)\")") || text.contains("UIImage(named: \"\(name)\")") {
                                    find.append(asset)
                                    fputs("\(file) find \(name)\n", stdout)
                                }
                                // else if text.range(of: "UIImage[\\.init]*\\(named: *\"\(name)\"", options: .regularExpression, range: nil, locale: nil) != nil {
                                //     find.append(asset)
                                //     fputs("\(file) find \(name)\n", stdout)
                                // }
                            }
                            for asset in find {
                                assets[x].imageset.remove(asset)
                            }
                        }
                    } catch {
                        fputs("\(error.localizedDescription)\n", stderr)
                    }
                }
            }
            fputs("\n\n\n\nunused imageset\n", stdout)
            let output = assets.map({ item -> [String] in
                return item.imageset.map({ image -> String in
                    return "\(item.xcassets) -> \(image.lastComponentWithoutExtension)"
                })
            }).flatMap({ $0 }).joined(separator: "\n")
            fputs(output, stdout)
        } catch {
            fputs("error: \(error.localizedDescription)\n\(help)", stderr)
            exit(1)
        }
    }
}

File: internal/asset/asset.go
package asset

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/TBXark/xcode-tool/internal/file"
)

// Contents represents the structure of an asset Contents.json file.
type Contents struct {
	Images []struct {
		Filename string `json:"filename"`
		Idiom    string `json:"idiom"`
		Scale    string `json:"scale"`
	} `json:"images"`
	Info struct {
		Author  string `json:"author"`
		Version int    `json:"version"`
	} `json:"info"`
}

// Group tracks the imagesets that belong to a .xcassets bundle.
type Group struct {
	XCAssets  string
	Imagesets map[string]bool
}

// Rename renames asset files in .imageset directories to match the asset name.
func Rename(dirPath string) error {
	paths, err := file.FindAllDirectoryPaths(dirPath, ".imageset")
	if err != nil {
		return fmt.Errorf("finding imageset directories: %w", err)
	}

	fmt.Printf("Find %d assets in %s\n", len(paths), dirPath)

	for _, assetPath := range paths {
		if err := renameAsset(assetPath); err != nil {
			log.Printf("Error processing %s: %v", assetPath, err)
		}
	}

	return nil
}

func renameAsset(assetPath string) error {
	jsonPath := filepath.Join(assetPath, "Contents.json")

	// Read JSON file
	data, err := os.ReadFile(jsonPath)
	if err != nil {
		return fmt.Errorf("reading %s: %w", jsonPath, err)
	}

	var contents Contents
	if err := json.Unmarshal(data, &contents); err != nil {
		return fmt.Errorf("parsing %s: %w", jsonPath, err)
	}

	assetName := file.GetFileNameWithoutExt(assetPath)
	didUpdate := false
	jsonStr := string(data)

	for _, image := range contents.Images {
		if image.Filename == "" {
			continue
		}

		parts := strings.Split(image.Filename, ".")
		if len(parts) < 2 {
			continue
		}

		ext := parts[len(parts)-1]
		scale := ""
		if image.Scale != "" && image.Scale != "1x" {
			scale = "@" + image.Scale
		}

		targetName := fmt.Sprintf("%s%s.%s", assetName, scale, ext)

		if targetName != image.Filename {
			oldPath := filepath.Join(assetPath, image.Filename)
			newPath := filepath.Join(assetPath, targetName)

			if err := os.Rename(oldPath, newPath); err != nil {
				log.Printf("Error renaming %s: %v", oldPath, err)
				continue
			}

			// Update JSON string
			pattern := fmt.Sprintf(`"filename"\s*:\s*"%s"`, regexp.QuoteMeta(image.Filename))
			re := regexp.MustCompile(pattern)
			jsonStr = re.ReplaceAllString(jsonStr, fmt.Sprintf(`"filename" : "%s"`, targetName))
			didUpdate = true
		}
	}

	if didUpdate {
		if err := os.WriteFile(jsonPath, []byte(jsonStr), 0644); err != nil {
			return fmt.Errorf("writing %s: %w", jsonPath, err)
		}
	}

	return nil
}

// Clean finds and reports unused image assets by scanning Swift files.
func Clean(dirPath string) error {
	// Find all .xcassets directories
	xcassetsPaths, err := file.FindAllDirectoryPaths(dirPath, ".xcassets")
	if err != nil {
		return fmt.Errorf("finding xcassets directories: %w", err)
	}

	// Map each .xcassets to its imagesets
	var assets []Group
	for _, xcassetsPath := range xcassetsPaths {
		imagesets, err := file.FindAllDirectoryPaths(xcassetsPath, ".imageset")
		if err != nil {
			log.Printf("Error: %v", err)
			continue
		}

		imagesetMap := make(map[string]bool)
		for _, imageset := range imagesets {
			imagesetMap[imageset] = true
		}

		assets = append(assets, Group{
			XCAssets:  xcassetsPath,
			Imagesets: imagesetMap,
		})
	}

	// Find all Swift files
	swiftFiles, err := file.FindAllFilePaths(dirPath, ".swift")
	if err != nil {
		return fmt.Errorf("finding swift files: %w", err)
	}

	// Filter out R.generated.swift files
	var filteredFiles []string
	for _, f := range swiftFiles {
		if filepath.Base(f) != "R.generated.swift" {
			filteredFiles = append(filteredFiles, f)
		}
	}

	// Scan Swift files for asset usage
	for _, f := range filteredFiles {
		content, err := os.ReadFile(f)
		if err != nil {
			log.Printf("Error reading %s: %v", f, err)
			continue
		}

		text := string(content)

		for i := range assets {
			var found []string
			for imageset := range assets[i].Imagesets {
				name := file.GetFileNameWithoutExt(imageset)

				// Check for various patterns
				if strings.Contains(text, fmt.Sprintf("R.image.%s", name)) ||
					strings.Contains(text, fmt.Sprintf(`#imageLiteral(resourceName: "%s")`, name)) ||
					strings.Contains(text, fmt.Sprintf(`UIImage(named: "%s")`, name)) {
					found = append(found, imageset)
					fmt.Printf("%s find %s\n", f, name)
				}
			}

			// Remove found assets
			for _, imageset := range found {
				delete(assets[i].Imagesets, imageset)
			}
		}
	}

	// Print unused assets
	fmt.Printf("\n\n\n\nunused imageset\n")
	for _, asset := range assets {
		for imageset := range asset.Imagesets {
			name := file.GetFileNameWithoutExt(imageset)
			fmt.Printf("%s -> %s\n", asset.XCAssets, name)
		}
	}

	return nil
}
